### ISSUE TITLE
[clarification]: requirement to support lifecycle methods returning wrapper types and success counts

### ISSUE THREAD

**Issue by gavinking** [jakartaee/data#488]
### Specification

Javadoc of @Insert and friends 

### I need clarification on ...

- Why support for wrapper types is even required.
- Whether lifecycle methods should return success counts instead of throwing.

### Additional information

In the Javadoc for `@Insert`, `@Update`, `@Delete`, and `@Save`, we've actually grown a bit of a combinatorial explosion of possible supported signatures.

We have to accept:

1. single enitities
2. arrays
3. `iterable`s

and we have to be able to return

1. `void`
2. `boolean`
3. `int` or `long`
4. `Boolean`, `Integer`, or `Long`
5. `Void`, in some cases
6. in the case of `@Insert`, an array 
7. in the case of `@Insert`, an `Iterable` 

Wrapper types and `Void`
---------

First, what is the point of accepting `Void` here? And why allow wrapper types? Just to increase the possibility of an NPE?

IMO, support for returning wrapper types should not be required.

"Quiet" failure
---------

The requirement to support return types other than `void` on a _lifecycle_ (as opposed to _query_) method came a bit out of the blue for me, since I would have said that such methods _should always fail with an exception_ if they can't perform the operation they've been asked to perform.

To me it seems like the operation:

```java
@Update
int update(Iterable<Book> book)
```

is really quite a dangerous API to allow, in that it's highly likely that a client will forget to check the return value of the method, in which case a failure goes unnoticed.

It's almost _always_ going to be the case that if I ask my repository to store 5 `Book`s at once, and it only succeeds in storing 4 of them, then what I want to do is fail and roll back the transaction. Indeed, being able to recover from such a failure _and just keep on working_ is something that's not necessarily even _possible_. At the very least, if I plan to keep working after such a failure, then I need to know _why_ it failed, and an `int` doesn't give me that information! Hell, very often the _database_ doesn't give me that information in a useful form.

So supporting this feels like it leads users down a bad path, and implementors down a rabbit hole of trying to understand which sorts of failures are recoverable.

I don't think this should be required in Jakarta Data 1.0.

Can we please simplify the requirements here, at least for this release?

**Comment by gavinking** [comment]
P.S. note that typically when you execute an `update` or `delete` and get a zero row count back, that's due to an optimistic lock failure and it should be reported via `OptimisticLockingFailureException`.

Violation of a database-level constraint should also always be reported via an exception, in my opinion.

I actually have no real clue when it would be appropriate to return `boolean`.

**Comment by gavinking** [comment]
Also, the current Javadoc for `@Update` is not even self-consistent. It currently says that:

> The return type of the annotated method must be {@code void}, {@code boolean}, {@code int}, {@code long} a corresponding primitive wrapper type such as {@link Integer}, or the same type as the parameter.

But then goes on to contradict itself by saying:

>  An {@code Iterable} or array return type includes all matching entities that are found in the database,  skipping over entities that are not present in the database or have a non-matching version.  For example, if the method is annotated with {@code @Update} and takes a parameter of type {@code Car car}, the return type can be {@code Car}.

**Comment by gavinking** [comment]
Uffffffff, I just noticed this prescriptive language:

> If the entity does not exist in the database or it is versioned and its version differs from the version in the database,  no update is made and no error is raised.

This is Bad, people, really Bad. 😱

**Comment by gavinking** [comment]
> Uffffffff, I just noticed this prescriptive language from `@Update`:

There's one sort of bug/problem which our team takes absolutely far more seriously than any other kind of bug: it's the bug which can lead to data being _silently_ corrupted at runtime. Not only should our product not have this sort of bug, but it should make it very difficult for the user to write programs which have this sort of bug. I strongly advocate that Jakarta Data take the same approach.

**Comment by njr-11** [comment]
> First, what is the point of accepting `Void` here? And why allow wrapper types? Just to increase the possibility of an NPE?

I don't know why `Void` and wrapper types are listed as required.  I can't think of any reason why a user would want `Void` and the wrapper types don't seem to add any value, and as you stated only add uncertainty around the possibility of a `null` value.  I would be fine with omitting these.

Regarding optimistic locking, I can think of different scenarios where `OptimisticLockingFailureException` vs a boolean false return value are desirable.  I do like how `OptimisticLockingFailureException` can be used to avoid users errors with optimistic logic where they forget to check the return value.  However, in scenarios involving transactions that span multiple operations, `OptimisticLockingFailureException` can wreck an optimistic locking approach because making the optimistic locking result of false into an exception can cause the entire transaction to need be rolled back, preventing the user from proceeding to the next part of their business logic that deals with the false value.  I would like to be able to support both approaches, based on which return type the user asks for, as the `Delete` annotation is currently doing.  In that case, a `void` return type means you want `OptimisticLockingFailureException` whereas a `boolean` return type means you want a `true`/`false` value to indicate whether it was deleted.

As to whether we want numeric return types, I think those have less value in version 1.0, but looking ahead I can see them becoming useful.  For example, imagine if, in the future, in order to attempt to match more of what you can do with Query by Method Name, we were to allow the same sort of pattern with `Delete` that we already have with `Find`, then a numeric return type will be useful:

```
@Delete
long delete(String make, String model, int year);
```

```
long howManyDeleted = vehicles.delete(make, model, year);
```

Given that this is something I would like to add in the future, I would be happy with keeping the numeric return values.

**Comment by gavinking** [comment]
>  However, in scenarios involving transactions that span multiple operations, `OptimisticLockingFailureException` can wreck an optimistic locking approach because making the optimistic locking result of false into an exception can cause the entire transaction to need be rolled back, preventing the user from proceeding to the next part of their business logic that deals with the false value.

Hibernate _always_ throws exceptions for optimistic failures for lifecycle method. I've never even once in > 20 years encountered a user who complained about this.

In the case where the user doesn't care about versioning they can, and _do_ use a `delete` or `update` query, but queries have quite different semantics and surrounding expectations to _lifecycle methods_.

If you want to implement and play with such experimental and frankly dangerous features in your implementation, that's fine, but it should not be required by Jakarta Data.

**Comment by gavinking** [comment]
> For example, imagine if, in the future, in order to attempt to match more of what you can do with Query by Method Name, we were to allow the same sort of pattern with Delete that we already have with Find, then a numeric return type will be useful.

Well, you're kind of mixing up queries with lifecycle methods here.

- Queries accept arbitrary parameters and can do arbitrary things.
- Lifecycle methods accept an _entity instance_ and do something with very clear semantics.

The capability you're describing here is already more than well covered by `@Query` methods. Now, if you want to reuse the `@Delete` annotation for _delete queries_, that's probably fine, but a delete query is still a very different thing to the delete operation which accepts an entity instance.

And that interpretation of `@Delete` is not in Jakarta Data 1.0.

**Comment by gavinking** [comment]
Just for the record: this requirement in principle makes it _impossible_ to implement Jakarta Data on top of JPA, since most exceptions emanating from JPA (including very explicitly the `OptimisticLockException`) mark the transaction for rollback.

**Comment by njr-11** [comment]
> Also, the current Javadoc for `@Update` is not even self-consistent. It currently says that:
> 
> > The return type of the annotated method must be {@code void}, {@code boolean}, {@code int}, {@code long} a corresponding primitive wrapper type such as {@link Integer}, or the same type as the parameter.
> 
> But then goes on to contradict itself by saying:
> 
> > An {@code Iterable} or array return type includes all matching entities that are found in the database,  skipping over entities that are not present in the database or have a non-matching version.  For example, if the method is annotated with {@code @update} and takes a parameter of type {@code Car car}, the return type can be {@code Car}.

I don't believe this is inconsistent.  The first statement includes that the return type can be the same type as the parameter.  Just prior to that it listed the allowable parameter types, which are
```
    * The entity to be updated.
    * An Iterable of entities to be updated.
    * An array of entities to be updated.
```

The second selection discusses those as return types.

That said, I did spot I place with an inconsistency right after that, where it states: "Similarly, if the parameter is an `Iterable<Car>` or an array of Car, the return type can be `Iterable<Car>`."

This is not consistent with the prior statement that requires the same type.  A correction could be:

That said, I did spot I place with an inconsistency right after that, where it states: "Similarly, if the parameter is an `Iterable<Car>`, the return type can be `Iterable<Car>`."

**Comment by njr-11** [comment]
> Now, if you want to reuse the `@Delete` annotation for _delete queries_, that's probably fine, but a delete query is still a very different thing to the delete operation which accepts an entity instance.
> 
> And that interpretation of `@Delete` is not in Jakarta Data 1.0.

What I'm struggling with here is how to be able to use the `Delete` annotation for both.  Even if we omit the numeric & boolean return type for 1.0, we would still be needing to add it back in vNext when attempting to use the `Delete` query approach as an alternative to Query by Method Name.  When I went through evaluating whether the life cycle annotations would be compatible with that approach, this is one of the aspects that made that the case.  I know one option would be to split the use cases between two different annotations, but I can't think of any way to meaningfully name them that users won't accidentally interchange them.  Maybe we can split the use cases based on the parameter types, with these life cycle entity-based parameter types requiring OptimisticLockingException?  I'll think about this some more and what could be done to make this clearer.  I don't want to change something here that inadvertently takes away our future direction.

**Comment by gavinking** [comment]
Let me rewind a bit, and explain myself better. I have no problem at all with this:

```java
@Query("delete from Book with id = ?1")
int deleteBook(String isbn);
```

That's clear an explicit and self-documenting. It's obvious what it does. It deletes by id and the version just isn't involved.

But now consider these method declarations:

```java
@Delete
void deleteBook(Book book);
```
```java
@Delete
int deleteBook(Book book);
```

One of these throws when something fails, one of them swallows the failure and returns 0.

Do you think this is very obvious to a caller of this API?

Now consider all of these:

```java
@Insert
void insert(Book book);

@Insert
void insert(Book[] books);

@Insert
Book insert(Book book);

@Insert
Book[] insert(Book[] books);

@Update
void update(Book book);

@Update
boolean update(Book book);

@Update
Book update(Book book);

@Update
Book[] update(Book book);

@Update
int update(Book book);
```

Which of these throw on failure and which of these swallow? I have read the Javadoc of the annotations quite closely and I'm still not entirely clear what is intended in all cases.

Do you think it's obvious to a _user_. (Remember that the caller of those methods might not be the person who wrote the repository, and might not be an expert on Jakarta Data.)

**Comment by gavinking** [comment]
> What I'm struggling with here is how to be able to use the Delete annotation for both. Even if we omit the numeric & boolean return type for 1.0, we would still be needing to add it back in vNext when attempting to use the Delete query approach as an alternative to Query by Method Name. 

Well there's a rather bright line here: in one case the parameter is of entity type, the the other case it's of primary key type.

I would not be surprised when these methods do different things:

```java
interface Library implements DataRepository<Book> {
    @Delete
    void deleteBook(Book book);
    
    @Delete
    boolean deleteBook(String isbn);
}
```

In one case I passed an entity instance _which has a version in it_, and in the other case I passed an id, and there's no version anywhere.

**Comment by gavinking** [comment]
This issue was fixed, after much productive discussion, by #490. Closing.
